File: .github/scripts/generate-changelog.py
#!/usr/bin/env python3
"""Generate a CHANGELOG.md entry from commits since the previous tag.

Usage: generate-changelog.py <version>
  e.g. generate-changelog.py v3.8

Outputs:
  - Prepends new section to CHANGELOG.md
  - Updates vault-manifest.json version and released date
  - Prints the generated section to stdout (for use as GitHub Release body)
"""

import json
import re
import subprocess
import sys
from datetime import datetime, timezone

PREFIX_MAP = {
    "feat": "Added",
    "fix": "Fixed",
    "docs": "Changed",
    "ci": "Changed",
    "refactor": "Changed",
    "perf": "Changed",
    "test": "Changed",
    "chore": "Changed",
    "build": "Changed",
    "style": "Changed",
    "revert": "Fixed",
}

SKIP_PREFIXES = {"release", "ci"}

SECTION_ORDER = ["Added", "Changed", "Fixed", "Removed"]


def get_previous_tag():
    result = subprocess.run(
        ["git", "tag", "--sort=-version:refname"],
        capture_output=True, text=True
    )
    tags = [t.strip() for t in result.stdout.strip().split("\n") if t.strip()]
    if len(tags) >= 2:
        return tags[1]
    return None


def get_commits(since_tag):
    if since_tag:
        range_spec = f"{since_tag}..HEAD"
    else:
        range_spec = "HEAD"

    result = subprocess.run(
        ["git", "log", range_spec, "--pretty=format:%s", "--first-parent"],
        capture_output=True, text=True
    )
    return [line.strip() for line in result.stdout.strip().split("\n") if line.strip()]


def classify_commit(message):
    # Strip PR reference suffix like (#25)
    clean = re.sub(r'\s*\(#\d+\)\s*$', '', message)

    # Try prefix match: "feat: description" or "feat(scope): description"
    match = re.match(r'^(\w+)(?:\([^)]*\))?\s*:\s*(.+)$', clean)
    if match:
        prefix = match.group(1).lower()
        description = match.group(2).strip()
        if prefix in SKIP_PREFIXES:
            return None, None
        category = PREFIX_MAP.get(prefix, "Changed")
        return category, description

    # No prefix — capitalize first letter, put in Changed
    if not clean:
        return "Changed", clean
    return "Changed", clean[0].upper() + clean[1:]


def generate_section(version, commits):
    today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    grouped = {}

    for msg in commits:
        category, description = classify_commit(msg)
        if category is None:
            continue
        if category not in grouped:
            grouped[category] = []
        grouped[category].append(description)

    lines = [f"## {version} — {today}", ""]

    for section in SECTION_ORDER:
        if section in grouped:
            lines.append(f"### {section}")
            for item in grouped[section]:
                lines.append(f"- {item}")
            lines.append("")

    return "\n".join(lines)


def prepend_to_changelog(section, version):
    with open("CHANGELOG.md", "r", encoding="utf-8") as f:
        content = f.read()

    # Insert after the "# Changelog" header
    header = "# Changelog"
    idx = content.find(header)
    if idx == -1:
        new_content = f"{header}\n\n{section}\n{content}"
    else:
        insert_at = idx + len(header)
        # Skip any whitespace after header
        while insert_at < len(content) and content[insert_at] in ("\n", "\r"):
            insert_at += 1
        new_content = content[:insert_at] + "\n" + section + "\n" + content[insert_at:]

    with open("CHANGELOG.md", "w", encoding="utf-8") as f:
        f.write(new_content)


def update_manifest(version):
    with open("vault-manifest.json", "r", encoding="utf-8") as f:
        manifest = json.load(f)

    # Strip 'v' prefix for semver
    semver = version.lstrip("v")
    manifest["version"] = semver
    manifest["released"] = datetime.now(timezone.utc).strftime("%Y-%m-%d")

    with open("vault-manifest.json", "w", encoding="utf-8") as f:
        json.dump(manifest, f, indent=2, ensure_ascii=False)
        f.write("\n")


def main():
    if len(sys.argv) < 2:
        print("Usage: generate-changelog.py <version>", file=sys.stderr)
        sys.exit(1)

    version = sys.argv[1]
    prev_tag = get_previous_tag()
    commits = get_commits(prev_tag)

    if not commits:
        print("No commits found since previous tag.", file=sys.stderr)
        sys.exit(1)

    section = generate_section(version, commits)

    prepend_to_changelog(section, version)
    update_manifest(version)

    # Print section for GitHub Release body
    print(section)


if __name__ == "__main__":
    main()

File: .github/scripts/generate-changelog.js
#!/usr/bin/env node
/**
 * Generate a CHANGELOG.md entry from commits since the previous tag.
 *
 * Usage: node generate-changelog.js <version>
 *   e.g. node generate-changelog.js v3.8
 *
 * Outputs:
 *   - Prepends new section to CHANGELOG.md (or replaces if version exists)
 *   - Updates vault-manifest.json version and released date
 *   - Prints the generated section to stdout (for use as GitHub Release body)
 */
'use strict';

const { execFileSync } = require('child_process');
const fs = require('fs');

const PREFIX_MAP = {
  feat: 'Added',
  fix: 'Fixed',
  docs: 'Changed',
  ci: 'Changed',
  refactor: 'Changed',
  perf: 'Changed',
  test: 'Changed',
  chore: 'Changed',
  build: 'Changed',
  style: 'Changed',
  revert: 'Fixed',
};

const SKIP_PREFIXES = new Set(['release', 'ci', 'test']);

const SECTION_ORDER = ['Added', 'Changed', 'Fixed', 'Removed'];

function runGit(...args) {
  try {
    return execFileSync('git', args, { encoding: 'utf-8' }).trim();
  } catch (err) {
    process.stderr.write(`git ${args.join(' ')} failed: ${err.stderr || err.message}\n`);
    process.exit(1);
  }
}

function getPreviousTag() {
  const output = runGit('tag', '--sort=-version:refname');
  const tags = output.split('\n').map((t) => t.trim()).filter(Boolean);
  if (tags.length >= 2) {
    return tags[1];
  }
  return null;
}

function getCommits(sinceTag) {
  const rangeSpec = sinceTag ? `${sinceTag}..HEAD` : 'HEAD';
  const output = runGit('log', rangeSpec, '--pretty=format:%s', '--first-parent');
  return output.split('\n').map((l) => l.trim()).filter(Boolean);
}

function classifyCommit(message) {
  // Strip PR reference suffix like (#25)
  const clean = message.replace(/\s*\(#\d+\)\s*$/, '');

  // Try prefix match: "feat: description" or "feat(scope): description"
  const match = clean.match(/^(\w+)(?:\([^)]*\))?\s*:\s*(.+)$/);
  if (match) {
    const prefix = match[1].toLowerCase();
    const description = match[2].trim();
    if (SKIP_PREFIXES.has(prefix)) {
      return [null, null];
    }
    const category = PREFIX_MAP[prefix] || 'Changed';
    return [category, description];
  }

  // No prefix — capitalize first letter, put in Changed
  if (!clean) {
    return ['Changed', clean];
  }
  return ['Changed', clean[0].toUpperCase() + clean.slice(1)];
}

function generateSection(version, commits) {
  const today = new Date().toISOString().slice(0, 10);
  const grouped = {};

  for (const msg of commits) {
    const [category, description] = classifyCommit(msg);
    if (category === null) continue;
    if (!grouped[category]) grouped[category] = [];
    grouped[category].push(description);
  }

  const lines = [`## ${version} \u2014 ${today}`, ''];

  for (const section of SECTION_ORDER) {
    if (grouped[section]) {
      lines.push(`### ${section}`);
      for (const item of grouped[section]) {
        lines.push(`- ${item}`);
      }
      lines.push('');
    }
  }

  return lines.join('\n');
}

function prependToChangelog(section, version) {
  let content = fs.readFileSync('CHANGELOG.md', 'utf-8');

  // Check if this version already exists (idempotent on re-runs)
  const versionPattern = new RegExp(`^## ${version.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')} \u2014 .*$`, 'm');
  const existingPattern = new RegExp(
    `^## ${version.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')} \u2014 .*?(?=\\n## [^\\n]|$)`,
    'ms'
  );

  let newContent;
  if (versionPattern.test(content)) {
    newContent = content.replace(existingPattern, section.trimEnd());
  } else {
    // Insert after the "# Changelog" header
    const header = '# Changelog';
    const idx = content.indexOf(header);
    if (idx === -1) {
      newContent = `${header}\n\n${section}\n${content}`;
    } else {
      let insertAt = idx + header.length;
      while (insertAt < content.length && (content[insertAt] === '\n' || content[insertAt] === '\r')) {
        insertAt++;
      }
      newContent = content.slice(0, insertAt) + '\n' + section + '\n' + content.slice(insertAt);
    }
  }

  fs.writeFileSync('CHANGELOG.md', newContent, 'utf-8');
}

function normalizeVersion(version) {
  const match = version.match(/^v?(\d+)\.(\d+)(?:\.(\d+))?$/);
  if (!match) {
    process.stderr.write(`Invalid version '${version}'. Expected vX.Y or vX.Y.Z.\n`);
    process.exit(1);
  }
  const [, major, minor, patch] = match;
  return `${major}.${minor}.${patch || '0'}`;
}

function updateManifest(version) {
  const manifest = JSON.parse(fs.readFileSync('vault-manifest.json', 'utf-8'));
  manifest.version = normalizeVersion(version);
  manifest.released = new Date().toISOString().slice(0, 10);
  fs.writeFileSync('vault-manifest.json', JSON.stringify(manifest, null, 2) + '\n', 'utf-8');
}

function main() {
  if (process.argv.length < 3) {
    process.stderr.write('Usage: node generate-changelog.js <version>\n');
    process.exit(1);
  }

  const version = process.argv[2];
  const prevTag = getPreviousTag();
  const commits = getCommits(prevTag);

  if (commits.length === 0) {
    process.stderr.write('No commits found since previous tag.\n');
    process.exit(1);
  }

  const section = generateSection(version, commits);

  if (!commits.some((msg) => classifyCommit(msg)[0] !== null)) {
    process.stderr.write('All commits were skipped (ci/test/release only). Nothing to changelog.\n');
    process.exit(1);
  }

  prependToChangelog(section, version);
  updateManifest(version);

  // Print section for GitHub Release body
  console.log(section);
}

main();
